# Практическое занятие 2
# Компьютерный практикум по алгебре на Python

## Матрицы. Подстановки, действия над строками, столбцами и элементами. Формулы Крамера для решения СЛАУ.

https://docs.sympy.org/latest/tutorial/matrices.html

С помощью конструктора класса Matrix можно создавать матрицы, элементами которых служат не только числа или числовые выражения, но и символы Symbol, играющие роль математических переменных. Переменные могут принимать какие-то значения, эти значения можно подставлять в матрицы (и не только в них!) с помощью метода subs.

In [4]:
from sympy import Symbol, symbols, Matrix, pi, evalf, sqrt, Eq

### Задание 1
Решить по формулам Крамера СЛАУ
\begin{align*}
\left\{\begin{matrix} 2 x_{1} +  x_{3}  = -8
\\- 4 x_{1} + 5 x_{2}  = 20
\\- 6 x_{2} + x_{3}  = 2
\end{matrix}\right.
 \end{align*}
Проверить правильность найденного решения подстановкой (в математическом смысле! subs не нужно использовать).

Под проверкой правильности понимается проверка выполнения равенства $AX=b$ или $AX-b=\bar 0$.

In [5]:
A = Matrix([[2, 0, 1], [-4, 5, 0], [0, -6, 1]])
b = Matrix([-8, 20, 2])
det_A = A.det()
X = [0 for i in range(A.shape[0])]

for i in range(3):
  Ai = A.copy()
  Ai.col_del(i)
  X[i] = Ai.col_insert(i, b).det() / det_A

X = Matrix(X)
display((X.T), (A * X - b).T)

Matrix([[-5, 0, 2]])

Matrix([[0, 0, 0]])

### Задание 2
Решить по формулам Крамера СЛАУ с параметром
\begin{align*}
\left\{\begin{matrix} a x_{1} + 2 x_{2}  = -7
\\- 4 x_{1} + 5 x_{2} + 3 x_{3} = 9
\\ 3 x_{1} + x_{3}  = 2
\end{matrix}\right.
 \end{align*}
Проверить правильность найденного решения подстановкой.

Составить матрицу $S$, столбцы которой являются решениями данной СЛАУ при $a = -1, 0, 3$.

**Указание.**

Оформить решение по формулам Крамера в виде функции.

Параметр $a$ описать как символ.

Использовать подстановки **subs** для построения матриц данной СЛАУ при $a = -1, 0, 3$ на основе матрицы с параметром.

Воспользоваться **simplify** для того, чтобы упростить выражение при проверке.

In [8]:
x = Symbol('a')
A = Matrix([[x, 2, 0], [-4, 5, 3], [3, 0, 1]])
b = Matrix([-7, 9, 2])
det_A = A.det()
len = [i for i in A.shape]

def func(var):
  X = [0 for i in range(len[0])]
  for i in range(len[0]):
    Si = A.copy()
    Si.col_del(i)
    X[i] = Si.col_insert(i, b).det() / det_A

  return (Matrix(X).subs(x, var))

a = [-1, 0, 3]
for i in range(3):
  print(f'При а = {a[i]}:\n')
  display(func(a[i]))
  print('\n')


При а = -1:



Matrix([
[-41/21],
[-94/21],
[  55/7]])



При а = 0:



Matrix([
[-41/26],
[  -7/2],
[175/26]])



При а = 3:



Matrix([
[-1],
[-2],
[ 5]])

### Задание 3.
Решить по формулам Крамера СЛАУ с параметрами
\begin{align*}
\left\{\begin{matrix} x_{1} + a_{12} x_{2}  = b_1
\\ x_{1} + a_{22} x_{2}  = b_2
\end{matrix}\right.
\end{align*}
Проверить правильность найденного решения подстановкой.

Найти такие наборы параметров, для которых полученное решение неприменимо (подсказка - из-за деления на ноль).

Показать, что для этих наборов определитель матрицы левой части равен нулю, использовать подстановки subs.

In [5]:
a12, a22 = symbols('a12 a22')
b1, b2 = symbols('b1 b2')
x1, x2 = symbols('x1 x2')

A = Matrix([[1, a12], [1, a22]])
b = Matrix([b1, b2])

X = Matrix([0, 0])
A_det = A.det()
for i in range(A.shape[0]):
    Ai = A.copy()
    Ai.col_del(i)
    X[i] = Ai.col_insert(i, b).det() / A_det
display(X.T)
display(Eq(A * X, b))
print('\nРешение неприменимо (знаменатель равен нулю) при (a22 - a12) = 0, т.е. при равенстве параметров, например 10:')
a12_p = a22_p = 10
display(X.T.subs([(a12, a12_p), (a22, a22_p)]))
print('\nОпределитель левой части для такого набора равен:')
display(A.subs([(a12, a12_p), (a22, a22_p)]).det())



Matrix([[(-a12*b2 + a22*b1)/(-a12 + a22), (-b1 + b2)/(-a12 + a22)]])

Eq(Matrix([
[a12*(-b1 + b2)/(-a12 + a22) + (-a12*b2 + a22*b1)/(-a12 + a22)],
[a22*(-b1 + b2)/(-a12 + a22) + (-a12*b2 + a22*b1)/(-a12 + a22)]]), Matrix([
[b1],
[b2]]))


Решение неприменимо (знаменатель равен нулю) при (a22 - a12) = 0, т.е. при равенстве параметров, например 10:


Matrix([[zoo*(10*b1 - 10*b2), zoo*(-b1 + b2)]])


Определитель левой части для такого набора равен:


0

### Задание 4.
Дана матрица
\begin{align*}
M =
\left(\begin{matrix}
p & 1 & 0 \\
1 & q & 1 \\
0 & 1 & r
\end{matrix}\right)
\end{align*}
Вычислить определитель этой матрицы.
Найти целочисленные наборы параметров $p$, $q$, $r$, для которых определитель $M$ равен нулю. Рассматривать целые числа от 0 до 12 (НЕ включительно). Использовать подстановки subs в заранее вычисленный определитель.

Из найденных наборов, содержащих равный 0 элемент, как столбцов составить матрицу $P$.

Из найденных наборов, состоящих только из положительных элементов, как строк составить матрицу $R$.

In [7]:
# Два последних предложения в условии написаны криво, поэтому не серчайте, если я неправильно его понял 😁.
p, q, r = symbols('p q r')
M = Matrix([[p, 1, 0], [1, q, 1], [0, 1, r]])
det_M = M.det()
P = []
R = []

for i in range(0, 12):
  for j in range(0, 12):
    for w in range(0, 12):
      d = det_M.subs({p: i, q: j, r: w})
      if d == 0:
        if (i == 0 or j == 0 or w == 0):
          P.append([i, j, w])
        else:
          R.append([i, j, w])

print('Matrix Р:\n')
display(Matrix(P).T)
print('\nMatrix R:\n')
display(Matrix(R))

Matrix Р:



Matrix([
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  0,  0],
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  0,  0]])


Matrix R:



Matrix([
[1, 2, 1],
[2, 1, 2]])

### Индивидуальное задание.
1) В матрице $A$ выполнить заданную подстановку.

2) В матрице $B$
выполнить подстановки в форме списка,
а также выполнить подстановки в форме словаря.

Подстановки в форме словаря выполнить одновременно и не одновременно.


Результат каждой подстановки выводить в виде отдельной матрицы.


Индивидуальное задание.

Вариант №175 (Просто скопировал из файла для удобства)


    1) В матрице
\begin{align*}
\left[\begin{matrix}9 & 6 & 16\\y & 8 & 7\\6 & 18 & 9\\6 & 4 & c\\1 & 3 & 9\\8 & 6 & 2\\17 & 7 & 10\end{matrix}\right]
\end{align*}
выполнить подстановку $c=\frac{7 y}{c}$


    2) В матрице
\begin{align*}
\left[\begin{matrix}v + 2 & 19 & 9 & 0 & 15 & 6 & 11\\14 & 18 & 16 & 17 & h + 4 & r + 3 & 1\\13 & 5 & 14 & 4 & 16 & 3 & 6\\4 & 3 & 13 & 15 & x - 4 & 4 & 19\\5 & 10 & 19 & 11 & 2 & q + 2 & 16\end{matrix}\right]
\end{align*}

выполнить подстановки в форме списка

$q=q - 2 r$,

$r=\frac{5 q}{v}$,

$v=q - 3 x$

а также выполнить следующие подстановки в форме словаря:

$q=v + 4$,

$v=q - 2$


    Подстановки в форме словаря выполнить одновременно и не одновременно


    Результат каждой подстановки выводить в виде отдельной матрицы

In [12]:
# Номер одын

y, c = symbols('y c')
d = 7 * y / c
matrix = Matrix([[9,  6,  16], \
                 [y,  8,  7], \
                 [6,  18, 9], \
                 [6,  4,  c], \
                 [1,  3,  9], \
                 [8,  6,  2], \
                 [17, 7,  10]])
matrix.subs(c, d)


Matrix([
[ 9,  6,    16],
[ y,  8,     7],
[ 6, 18,     9],
[ 6,  4, 7*y/c],
[ 1,  3,     9],
[ 8,  6,     2],
[17,  7,    10]])

In [28]:
# Номер два

q, r, v, h, x = symbols('q r v h x')
matrix = Matrix([[v + 2,  19, 9,  0,  15,    6,     11], \
                 [14,     18, 16, 17, h + 4, r + 3, 1 ], \
                 [13,     5,  14, 4,  16,    3,     6 ], \
                 [4,      3,  13, 15, x - 4, 4,     19], \
                 [5,      10, 19, 11, 2,     q + 2, 16]])

print('Подстановка в форме списка:\n')
display(matrix.subs([(q, q - 2 * r), (r, 5 * q / v), (v, q - 3 * x)]))

print('\nПодстановка в форме словаря не одновременная:\n')
display(matrix.subs({q: v + 4, v: q - 2}))

print('\nПодстановка в форме словаря одновременная:\n')
display(matrix.subs({q: v + 4, v: q - 2}, simultaneous=True))


Подстановка в форме списка:



Matrix([
[q - 3*x + 2, 19,  9,  0,    15,                      6, 11],
[         14, 18, 16, 17, h + 4,      5*q/(q - 3*x) + 3,  1],
[         13,  5, 14,  4,    16,                      3,  6],
[          4,  3, 13, 15, x - 4,                      4, 19],
[          5, 10, 19, 11,     2, q - 10*q/(q - 3*x) + 2, 16]])


Подстановка в форме словаря не одновременная:



Matrix([
[ q, 19,  9,  0,    15,     6, 11],
[14, 18, 16, 17, h + 4, r + 3,  1],
[13,  5, 14,  4,    16,     3,  6],
[ 4,  3, 13, 15, x - 4,     4, 19],
[ 5, 10, 19, 11,     2, q + 4, 16]])


Подстановка в форме словаря одновременная:



Matrix([
[ q, 19,  9,  0,    15,     6, 11],
[14, 18, 16, 17, h + 4, r + 3,  1],
[13,  5, 14,  4,    16,     3,  6],
[ 4,  3, 13, 15, x - 4,     4, 19],
[ 5, 10, 19, 11,     2, v + 6, 16]])